engine/consensus/matching/core.go

```diff
@@ -541,243 +539,265 @@ func (c *Core) CheckSealing() error {
+			return nil, nil, fmt.Errorf("internal error matching chunk approvals to incorporated result: %w", err)
 		}
+		sealableWithEnoughApprovals := sealingStatus.SufficientApprovalsForSealing
 
-		// The production system must always have a system chunk. Any result without chunks
-		// is wrong and should _not_ be sealed. While it is fundamentally the ReceiptValidator's
-		// responsibility to filter out results without chunks, there is no harm in also in enforcing
-		// the same condition here as well. It simplifies the code, because otherwise the
-		// matching core must enforce equality of start and end state for a result with zero chunks,
-		// in the absence of anyone else doing do.
-		matched := false
-		unmatchedIndex := -1
-		// check that each chunk collects enough approvals
-		for i, chunk := range incorporatedResult.Result.Chunks {
-			matched, err = c.matchChunk(incorporatedResult, block, chunk, assignment)
+		// Emergency Sealing Fallback: only kicks in if we can't seal following the happy-path sealing
+		emergencySealable := false
+		if !sealableWithEnoughApprovals {
+			emergencySealable, err = c.emergencySealable(incorporatedResult, lastFinalized)
 			if err != nil {
-				return nil, nil, fmt.Errorf("could not match chunk: %w", err)
-			}
-			if !matched {
-				unmatchedIndex = i
-				break
+				return nil, nil, fmt.Errorf("internal error matching chunk approvals to incorporated result: %w", err)
 			}
+			sealingStatus.SetQualifiesForEmergencySealing(emergencySealable)
 		}
 
-		emergencySealed := false
-		// ATTENTION: this is a temporary solution called emergency sealing. Emergency sealing is a special case
-		// when we seal ERs that don't have enough approvals but are deep enough in the chain resulting in halting sealing
-		// process. This will be removed when implementation of seal & verification is finished.
-		if !matched && c.emergencySealingActive {
-			incorporatedBlock, err := c.headersDB.ByBlockID(incorporatedResult.IncorporatedBlockID)
-			if err != nil {
-				return nil, nil, fmt.Errorf("could not get block %v: %w", incorporatedResult.IncorporatedBlockID, err)
-			}
-			// Note:
-			// we assume the incorporatedBlock is for a unsealed block, if
-			// there are DefaultEmergencySealingThreshold number of blocks between incorporatedBlock
-			// and lastFinalized, it means there are as many unsealed and finalized blocks, because
-			// should trigger the emergency sealing
-			if incorporatedBlock.Height+DefaultEmergencySealingThreshold <= lastFinalized.Height {
-				emergencySealed = true
+		// Determine sealability:
+		//  (i) the incorporatedResult must qualify for happy path sealing
+		//      or qualify for emergency sealing
+		// AND
+		// (ii) there must be at least 2 receipts from _different_ ENs
+		//      committing to the result
+		// comment: we evaluate condition (ii) only if (i) is true
+		if sealableWithEnoughApprovals || emergencySealable {
```

Comment:
[optional] there are quite a few nested `if` and `||` conditions in this function, would be great to keep the logic flat, however I haven't found a cleaner way. Maybe something for later in the next phase.

Reply:
tried to simplify: https://github.com/onflow/flow-go/blob/c72ae1ef036fad98d858d7fefa3c0b7ac8398d01/engine/consensus/matching/core.go#L598-L613

not sure whether this is simpler or just as complex (only in a different way) 😅